docs/userguide/ui-concepts.md

```diff
@@ -1,3 +1,151 @@
+
+=== "iOS"
+    ```Swift
+    struct WelcomeScreen: Screen {
+        var name: String
+        var onNameChanged: (String) -> Void
+        var onLoginTapped: () -> Void
+
+        func viewControllerDescription(environment: ViewEnvironment) -> ViewControllerDescription {
+            return WelcomeViewController.description(for: self, environment: environment)
+        }
+    }
+
+    private final class WelcomeViewController: ScreenViewController<WelcomeScreen> {
+        // ...
+    ```
+
+=== "Android Classic"
+    ```Kotlin
+    data class HelloScreen(
+      val message: String,
+      val onClick: () -> Unit
+    ) : AndroidScreen<HelloScreen> {
+      override val viewFactory: ScreenViewFactory<HelloScreen> =
+        fromViewBinding(HelloViewBinding::inflate) { r, _ ->
+          helloMessage.text = r.message
+          helloMessage.setOnClickListener { r.onClick() }
+        }
+    }
+    ```
+
+=== "Android Compose"
+    ```Kotlin
+    data class HelloScreen(
+      val message: String,
+      val onClick: () -> Unit
+    ) : AndroidScreen<HelloScreen> by ComposeScreen {
+      Button(onClick = onClick)
+      Text("message")
+    }
+    ```
+
+After all the chest-thumping above about [Separation of Concerns](#separation-of-concerns), these code samples probably don't look very separate. That's because, while the libraries themselves are completely decoupled, they don't force that strict rigor on your app code.
+
+If you aren't building, say, a core workflow module that you want to ship separately from its Android app, and an alternative command line interface, you'd probably gain nothing from enforced separation but boilerplate and runtime errors.
+And in practice, your Workflow unit tests won't call `viewControllerDescription()` or `viewFactory` and will build and run just fine against the JVM.
+
+If you're one of the few who really and trully do need impermeable boundaries between your core and UI modules, that's not hard to get. Your `Screen` implementations can be defined completely separately from their view code and bound later.
+
+
+=== "iOS"
+    ```Swift
+    // WelcomeScreen.swift
+    struct WelcomeScreen {
+        var name: String
+        var onNameChanged: (String) -> Void
+        var onLoginTapped: () -> Void
+    }
+
+    // WelcomeViewController.swift
+    extension WelcomeScreen: Screen {
+        func viewControllerDescription(environment: ViewEnvironment) -> ViewControllerDescription {
+            return WelcomeViewController.description(for: self, environment: environment)
+        }
+    }
+
+    private final class WelcomeViewController: ScreenViewController<WelcomeScreen> {
+        // ...
+    ```
```

Comment:
You can declare the struct with it implementing Screen and provide the implementation in an extension but it would have to be in the same swift module.